## <b>■ RNN 복습</b>
    2장. 개선된 통계기법을 이용해서 컴퓨터가 단어의 의미를 이해할 수 있또록 밀집벡터를 생성
    3장. CBOW 신경망을 만들어서 컴퓨터가 단어의 의미를 이해할 수 있도록 밀집벡터를 생성
    4장. CBOW 신경망에 큰 말뭉치가 들어갈 수 있도록 개선
        CBOW 신경망의 한계점
            1. 메모리 사용량이 많다
            2. 성능이 느리다
    5장. RNN 신경망 (기억하는 신경망)
        RNN 신경망의 한계점 : 장기 기억이 취약            
            1. 기울기 소실
            2. 기울기 폭발
    6장. LSTM 신경망
        LSTM 신경망과 RNN 신경망 차이점 : 기억셀
            1. forget 연산
            2. input/output 연산(기억 연산)
                새로 들어오는 데이터가 정보로써 얼마나 가치가 큰지 판단하는 게이트
        
        실습
            1. 긍정단어/부정단어를 컴퓨터가 인식할 수 있도록 LSTM 신경망 구성
                imdb (미국 영화 사이트 리뷰, 평점)

#### 문제1. 1부터 10까지 출력하시오 !

In [2]:
import numpy as np
np.arange(1,11)

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

#### 문제2. 60부터 940까지 출력하시오 ! 

In [3]:
np.arange(60,941)

array([ 60,  61,  62,  63,  64,  65,  66,  67,  68,  69,  70,  71,  72,
        73,  74,  75,  76,  77,  78,  79,  80,  81,  82,  83,  84,  85,
        86,  87,  88,  89,  90,  91,  92,  93,  94,  95,  96,  97,  98,
        99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111,
       112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124,
       125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137,
       138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150,
       151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163,
       164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176,
       177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189,
       190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202,
       203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215,
       216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228,
       229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 24

#### 문제3. 0번부터 120번까지 출력하는데 시계열 데이터를 만들기 위해서 아래와 같이 번호를 0번부터 60번씩 끊어서 출력하시오 !
    #결과 예시:
    0 60
    1  61
    2  62
    3  63
      :  
    57 117
    58 118
    59 119

In [5]:
for i in range(60,121):
    print(f"{i-60} {i}")

0 60
1 61
2 62
3 63
4 64
5 65
6 66
7 67
8 68
9 69
10 70
11 71
12 72
13 73
14 74
15 75
16 76
17 77
18 78
19 79
20 80
21 81
22 82
23 83
24 84
25 85
26 86
27 87
28 88
29 89
30 90
31 91
32 92
33 93
34 94
35 95
36 96
37 97
38 98
39 99
40 100
41 101
42 102
43 103
44 104
45 105
46 106
47 107
48 108
49 109
50 110
51 111
52 112
53 113
54 114
55 115
56 116
57 117
58 118
59 119
60 120


#### 문제4. 0번부터 940번까지 출력하는데 시계열 데이터를 만들기 위해서 아래와 같이 번호를 0번부터 60번씩 끊어서 출력하시오 !
    #결과 예시:
    0 60
    1 61
    2 62
    3 63
     :  
    878 938
    879 939
    880 940

In [7]:
for i in range(60,941):
    print(f"{i-60} {i}")

0 60
1 61
2 62
3 63
4 64
5 65
6 66
7 67
8 68
9 69
10 70
11 71
12 72
13 73
14 74
15 75
16 76
17 77
18 78
19 79
20 80
21 81
22 82
23 83
24 84
25 85
26 86
27 87
28 88
29 89
30 90
31 91
32 92
33 93
34 94
35 95
36 96
37 97
38 98
39 99
40 100
41 101
42 102
43 103
44 104
45 105
46 106
47 107
48 108
49 109
50 110
51 111
52 112
53 113
54 114
55 115
56 116
57 117
58 118
59 119
60 120
61 121
62 122
63 123
64 124
65 125
66 126
67 127
68 128
69 129
70 130
71 131
72 132
73 133
74 134
75 135
76 136
77 137
78 138
79 139
80 140
81 141
82 142
83 143
84 144
85 145
86 146
87 147
88 148
89 149
90 150
91 151
92 152
93 153
94 154
95 155
96 156
97 157
98 158
99 159
100 160
101 161
102 162
103 163
104 164
105 165
106 166
107 167
108 168
109 169
110 170
111 171
112 172
113 173
114 174
115 175
116 176
117 177
118 178
119 179
120 180
121 181
122 182
123 183
124 184
125 185
126 186
127 187
128 188
129 189
130 190
131 191
132 192
133 193
134 194
135 195
136 196
137 197
138 198
139 199
140 200
141 201
142 202
143 20

#### 문제5. 0부터 4704번까지 숫자로 이루어진 941행의 5열의 행렬을 만드시오

In [14]:
test_arr = np.arange(4705).reshape(-1,5)
print(test_arr[0])

[0 1 2 3 4]


#### 문제6. test_arr 행렬의 데이터에서 아래의 번호를 index로 해서 데이터를 추출하시오
    결과예시
    0 60
        [[  0   1   2   3   4]
         [  5   6   7   8   9]
         [ 10  11  12  13  14]
         [ 15  16  17  18  19]
         [ 20  21  22  23  24]
         [ 25  26  27  28  29]
         [ 30  31  32  33  34]
         [ 35  36  37  38  39]
         [ 40  41  42  43  44]
         [ 45  46  47  48  49]
         [ 50  51  52  53  54]
         [ 55  56  57  58  59]
         [ 60  61  62  63  64]
         [ 65  66  67  68  69]
         [ 70  71  72  73  74]
         [ 75  76  77  78  79]
         [ 80  81  82  83  84]
         [ 85  86  87  88  89]
         [ 90  91  92  93  94]
         [ 95  96  97  98  99]
         [100 101 102 103 104]
         [105 106 107 108 109]
         [110 111 112 113 114]
         [115 116 117 118 119]
         [120 121 122 123 124]
         [125 126 127 128 129]
         [130 131 132 133 134]
         [135 136 137 138 139]
         [140 141 142 143 144]
         [145 146 147 148 149]
         [150 151 152 153 154]
         [155 156 157 158 159]
         [160 161 162 163 164]
         [165 166 167 168 169]
         [170 171 172 173 174]
         [175 176 177 178 179]
         [180 181 182 183 184]
         [185 186 187 188 189]
         [190 191 192 193 194]
         [195 196 197 198 199]
         [200 201 202 203 204]
         [205 206 207 208 209]
         [210 211 212 213 214]
         [215 216 217 218 219]
         [220 221 222 223 224]
         [225 226 227 228 229]
         [230 231 232 233 234]
         [235 236 237 238 239]
         [240 241 242 243 244]
         [245 246 247 248 249]
         [250 251 252 253 254]
         [255 256 257 258 259]
         [260 261 262 263 264]
         [265 266 267 268 269]
         [270 271 272 273 274]
         [275 276 277 278 279]
         [280 281 282 283 284]
         [285 286 287 288 289]
         [290 291 292 293 294]
         [295 296 297 298 299]]

In [15]:
for i in range(60,62):
    print(f"{i-60} {i}")
    print(test_arr[i-60:i])

0 60
[[  0   1   2   3   4]
 [  5   6   7   8   9]
 [ 10  11  12  13  14]
 [ 15  16  17  18  19]
 [ 20  21  22  23  24]
 [ 25  26  27  28  29]
 [ 30  31  32  33  34]
 [ 35  36  37  38  39]
 [ 40  41  42  43  44]
 [ 45  46  47  48  49]
 [ 50  51  52  53  54]
 [ 55  56  57  58  59]
 [ 60  61  62  63  64]
 [ 65  66  67  68  69]
 [ 70  71  72  73  74]
 [ 75  76  77  78  79]
 [ 80  81  82  83  84]
 [ 85  86  87  88  89]
 [ 90  91  92  93  94]
 [ 95  96  97  98  99]
 [100 101 102 103 104]
 [105 106 107 108 109]
 [110 111 112 113 114]
 [115 116 117 118 119]
 [120 121 122 123 124]
 [125 126 127 128 129]
 [130 131 132 133 134]
 [135 136 137 138 139]
 [140 141 142 143 144]
 [145 146 147 148 149]
 [150 151 152 153 154]
 [155 156 157 158 159]
 [160 161 162 163 164]
 [165 166 167 168 169]
 [170 171 172 173 174]
 [175 176 177 178 179]
 [180 181 182 183 184]
 [185 186 187 188 189]
 [190 191 192 193 194]
 [195 196 197 198 199]
 [200 201 202 203 204]
 [205 206 207 208 209]
 [210 211 212 213 214]
 [215 

#### 문제7. 훈련 데이터와 훈련 데이터의 라벨을 같이 출력하시오

In [16]:
for i in range(60,941):
    print(test_arr[i-60:i])
    print(test_arr[i,0]) # 예측해야하는 정답

[[  0   1   2   3   4]
 [  5   6   7   8   9]
 [ 10  11  12  13  14]
 [ 15  16  17  18  19]
 [ 20  21  22  23  24]
 [ 25  26  27  28  29]
 [ 30  31  32  33  34]
 [ 35  36  37  38  39]
 [ 40  41  42  43  44]
 [ 45  46  47  48  49]
 [ 50  51  52  53  54]
 [ 55  56  57  58  59]
 [ 60  61  62  63  64]
 [ 65  66  67  68  69]
 [ 70  71  72  73  74]
 [ 75  76  77  78  79]
 [ 80  81  82  83  84]
 [ 85  86  87  88  89]
 [ 90  91  92  93  94]
 [ 95  96  97  98  99]
 [100 101 102 103 104]
 [105 106 107 108 109]
 [110 111 112 113 114]
 [115 116 117 118 119]
 [120 121 122 123 124]
 [125 126 127 128 129]
 [130 131 132 133 134]
 [135 136 137 138 139]
 [140 141 142 143 144]
 [145 146 147 148 149]
 [150 151 152 153 154]
 [155 156 157 158 159]
 [160 161 162 163 164]
 [165 166 167 168 169]
 [170 171 172 173 174]
 [175 176 177 178 179]
 [180 181 182 183 184]
 [185 186 187 188 189]
 [190 191 192 193 194]
 [195 196 197 198 199]
 [200 201 202 203 204]
 [205 206 207 208 209]
 [210 211 212 213 214]
 [215 216 2

 [1790 1791 1792 1793 1794]]
1795
[[1500 1501 1502 1503 1504]
 [1505 1506 1507 1508 1509]
 [1510 1511 1512 1513 1514]
 [1515 1516 1517 1518 1519]
 [1520 1521 1522 1523 1524]
 [1525 1526 1527 1528 1529]
 [1530 1531 1532 1533 1534]
 [1535 1536 1537 1538 1539]
 [1540 1541 1542 1543 1544]
 [1545 1546 1547 1548 1549]
 [1550 1551 1552 1553 1554]
 [1555 1556 1557 1558 1559]
 [1560 1561 1562 1563 1564]
 [1565 1566 1567 1568 1569]
 [1570 1571 1572 1573 1574]
 [1575 1576 1577 1578 1579]
 [1580 1581 1582 1583 1584]
 [1585 1586 1587 1588 1589]
 [1590 1591 1592 1593 1594]
 [1595 1596 1597 1598 1599]
 [1600 1601 1602 1603 1604]
 [1605 1606 1607 1608 1609]
 [1610 1611 1612 1613 1614]
 [1615 1616 1617 1618 1619]
 [1620 1621 1622 1623 1624]
 [1625 1626 1627 1628 1629]
 [1630 1631 1632 1633 1634]
 [1635 1636 1637 1638 1639]
 [1640 1641 1642 1643 1644]
 [1645 1646 1647 1648 1649]
 [1650 1651 1652 1653 1654]
 [1655 1656 1657 1658 1659]
 [1660 1661 1662 1663 1664]
 [1665 1666 1667 1668 1669]
 [1670 1671 16

 [3115 3116 3117 3118 3119]]
3120
[[2825 2826 2827 2828 2829]
 [2830 2831 2832 2833 2834]
 [2835 2836 2837 2838 2839]
 [2840 2841 2842 2843 2844]
 [2845 2846 2847 2848 2849]
 [2850 2851 2852 2853 2854]
 [2855 2856 2857 2858 2859]
 [2860 2861 2862 2863 2864]
 [2865 2866 2867 2868 2869]
 [2870 2871 2872 2873 2874]
 [2875 2876 2877 2878 2879]
 [2880 2881 2882 2883 2884]
 [2885 2886 2887 2888 2889]
 [2890 2891 2892 2893 2894]
 [2895 2896 2897 2898 2899]
 [2900 2901 2902 2903 2904]
 [2905 2906 2907 2908 2909]
 [2910 2911 2912 2913 2914]
 [2915 2916 2917 2918 2919]
 [2920 2921 2922 2923 2924]
 [2925 2926 2927 2928 2929]
 [2930 2931 2932 2933 2934]
 [2935 2936 2937 2938 2939]
 [2940 2941 2942 2943 2944]
 [2945 2946 2947 2948 2949]
 [2950 2951 2952 2953 2954]
 [2955 2956 2957 2958 2959]
 [2960 2961 2962 2963 2964]
 [2965 2966 2967 2968 2969]
 [2970 2971 2972 2973 2974]
 [2975 2976 2977 2978 2979]
 [2980 2981 2982 2983 2984]
 [2985 2986 2987 2988 2989]
 [2990 2991 2992 2993 2994]
 [2995 2996 29

 [4485 4486 4487 4488 4489]]
4490
[[4195 4196 4197 4198 4199]
 [4200 4201 4202 4203 4204]
 [4205 4206 4207 4208 4209]
 [4210 4211 4212 4213 4214]
 [4215 4216 4217 4218 4219]
 [4220 4221 4222 4223 4224]
 [4225 4226 4227 4228 4229]
 [4230 4231 4232 4233 4234]
 [4235 4236 4237 4238 4239]
 [4240 4241 4242 4243 4244]
 [4245 4246 4247 4248 4249]
 [4250 4251 4252 4253 4254]
 [4255 4256 4257 4258 4259]
 [4260 4261 4262 4263 4264]
 [4265 4266 4267 4268 4269]
 [4270 4271 4272 4273 4274]
 [4275 4276 4277 4278 4279]
 [4280 4281 4282 4283 4284]
 [4285 4286 4287 4288 4289]
 [4290 4291 4292 4293 4294]
 [4295 4296 4297 4298 4299]
 [4300 4301 4302 4303 4304]
 [4305 4306 4307 4308 4309]
 [4310 4311 4312 4313 4314]
 [4315 4316 4317 4318 4319]
 [4320 4321 4322 4323 4324]
 [4325 4326 4327 4328 4329]
 [4330 4331 4332 4333 4334]
 [4335 4336 4337 4338 4339]
 [4340 4341 4342 4343 4344]
 [4345 4346 4347 4348 4349]
 [4350 4351 4352 4353 4354]
 [4355 4356 4357 4358 4359]
 [4360 4361 4362 4363 4364]
 [4365 4366 43

    60일치의 데이터가 훈련 데이터로 출력되고 61번째의 데이터중 시가가 라벨로 출력

#### 문제8. 위의 훈련 데이터와 훈련 데이터의 라벨을 x_train과 y_train리스트에 담기게 하시오

In [18]:
x_train = []
y_train = []
for i in range(60,test_arr.shape[0]):
    x_train.append(test_arr[i-60:i])
    y_train.append(test_arr[i,0])
    
print(x_train[0])
print(y_train[0])

[[  0   1   2   3   4]
 [  5   6   7   8   9]
 [ 10  11  12  13  14]
 [ 15  16  17  18  19]
 [ 20  21  22  23  24]
 [ 25  26  27  28  29]
 [ 30  31  32  33  34]
 [ 35  36  37  38  39]
 [ 40  41  42  43  44]
 [ 45  46  47  48  49]
 [ 50  51  52  53  54]
 [ 55  56  57  58  59]
 [ 60  61  62  63  64]
 [ 65  66  67  68  69]
 [ 70  71  72  73  74]
 [ 75  76  77  78  79]
 [ 80  81  82  83  84]
 [ 85  86  87  88  89]
 [ 90  91  92  93  94]
 [ 95  96  97  98  99]
 [100 101 102 103 104]
 [105 106 107 108 109]
 [110 111 112 113 114]
 [115 116 117 118 119]
 [120 121 122 123 124]
 [125 126 127 128 129]
 [130 131 132 133 134]
 [135 136 137 138 139]
 [140 141 142 143 144]
 [145 146 147 148 149]
 [150 151 152 153 154]
 [155 156 157 158 159]
 [160 161 162 163 164]
 [165 166 167 168 169]
 [170 171 172 173 174]
 [175 176 177 178 179]
 [180 181 182 183 184]
 [185 186 187 188 189]
 [190 191 192 193 194]
 [195 196 197 198 199]
 [200 201 202 203 204]
 [205 206 207 208 209]
 [210 211 212 213 214]
 [215 216 2

#### 문제9. 위에서 append 시킨 x_train과 y_train에 무엇이 있는지 확인하시오

In [20]:
x_train[0], y_train[0]

(array([[  0,   1,   2,   3,   4],
        [  5,   6,   7,   8,   9],
        [ 10,  11,  12,  13,  14],
        [ 15,  16,  17,  18,  19],
        [ 20,  21,  22,  23,  24],
        [ 25,  26,  27,  28,  29],
        [ 30,  31,  32,  33,  34],
        [ 35,  36,  37,  38,  39],
        [ 40,  41,  42,  43,  44],
        [ 45,  46,  47,  48,  49],
        [ 50,  51,  52,  53,  54],
        [ 55,  56,  57,  58,  59],
        [ 60,  61,  62,  63,  64],
        [ 65,  66,  67,  68,  69],
        [ 70,  71,  72,  73,  74],
        [ 75,  76,  77,  78,  79],
        [ 80,  81,  82,  83,  84],
        [ 85,  86,  87,  88,  89],
        [ 90,  91,  92,  93,  94],
        [ 95,  96,  97,  98,  99],
        [100, 101, 102, 103, 104],
        [105, 106, 107, 108, 109],
        [110, 111, 112, 113, 114],
        [115, 116, 117, 118, 119],
        [120, 121, 122, 123, 124],
        [125, 126, 127, 128, 129],
        [130, 131, 132, 133, 134],
        [135, 136, 137, 138, 139],
        [140, 141, 1

#### 점심시간 문제: 위에는 시가 데이터를 라벨 정답으로 만들었지만 민혁이가 추천한 종가로 정답이 만들어지게 하시오 

In [ ]:
import tensorflow as tf
from tensorflow.keras.layers import Dense, LSTM, Dropout
from tensorflow.keras import Sequential
import numpy as np
import datetime
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.preprocessing import MinMaxScaler

# 1. 주가 데이터를 로드 합니다.
data = pd.read_csv("D:/Desktop/Itwill ws/rnn/goog.csv")

# 2. 
data['Date'] = pd.to_datetime(data['Date'])
split_date = datetime.datetime(2009, 1, 1)
training_data = data[data['Date'] < split_date].copy()
test_data = data[data['Date'] >= split_date].copy()

# 3.
training_data = training_data.drop(['Sym', 'Date', 'Adj Close'], axis=1)
training_data.head()

# 4. 정규화 
scaler = MinMaxScaler()
training_data = scaler.fit_transform(training_data)

# 5.
x_train = []
y_train = []

for i in range(60, training_data.shape[0]):
    x_train.append(training_data[i-60:i])
    y_train.append(training_data[i, 3])

print(x_train)
print(y_train)

####  1. 한국거래소(krx)에서 종목코드 가져온다.  상위 5개의 종목만 가져온다.

In [28]:
import pandas as pd 

code_df = pd.read_html('http://kind.krx.co.kr/corpgeneral/corpList.do?method=download&searchType=13', header=0)[0] 
code_df.종목코드 = code_df.종목코드.map('{:06d}'.format) # 종목코드가 6자리이기 때문에 6자리를 맞춰주기 위해 설정해줌 
code_df = code_df[['회사명', '종목코드']] # 우리가 필요한 것은 회사명과 종목코드이기 때문에 필요없는 column들은 제외해준다. 
code_df = code_df.rename(columns={'회사명': 'name', '종목코드': 'code'}) # 한글로된 컬럼명을 영어로 바꿔준다. 

print(code_df.head())

        name    code
0        DSR  155660
1         GS  078930
2      GS글로벌  001250
3  HDC현대산업개발  294870
4        KEC  092220


In [29]:
# 종목 이름을 입력하면 종목에 해당하는 코드를 불러와
# 네이버 금융(http://finance.naver.com)에 넣어줌

def get_url(item_name, code_df):
    code = code_df.query("name=='{}'".format(item_name))['code'].to_string(index=False).strip()
    url = 'http://finance.naver.com/item/sise_day.nhn?code={code}'.format(code=code)
    print("요청 URL = {}".format(url))

    return url

# GS의 일자데이터 url 가져오기

item_name='GS'
url = get_url(item_name,code_df)

# 일자 데이터를 담을 df라는 DataFrame 정의
df = pd.DataFrame()

# 1페이지에서 20페이지의 데이터만 가져오기
for page in range(1, 125):
    pg_url = '{url}&page={page}'.format(url=url, page=page)
    df = df.append(pd.read_html(pg_url, header=0)[0], ignore_index=True)

df = df.dropna()
print(df.head())

df.to_csv('D:/Desktop/Itwill ws/rnn/gs.csv', mode='w')

요청 URL = http://finance.naver.com/item/sise_day.nhn?code=035720
           날짜        종가      전일비        시가        고가        저가        거래량
1  2020.09.07  392000.0  10000.0  401500.0  401500.0  391000.0   896330.0
2  2020.09.04  402000.0   8000.0  390000.0  404500.0  389500.0  1168529.0
3  2020.09.03  410000.0   2000.0  415500.0  416000.0  404000.0   739215.0
4  2020.09.02  412000.0  10500.0  405500.0  412500.0  401500.0   921655.0
5  2020.09.01  401500.0   5500.0  413500.0  416000.0  400000.0  1100736.0


### ※ 문제. 본인이 주가 예측하기 원하는 회사의 주가 데이터 5년치를 csv 파일로 내려 받으시오

In [31]:
import pandas as pd 

code_df = pd.read_html('http://kind.krx.co.kr/corpgeneral/corpList.do?method=download&searchType=13', header=0)[0] 
code_df.종목코드 = code_df.종목코드.map('{:06d}'.format) # 종목코드가 6자리이기 때문에 6자리를 맞춰주기 위해 설정해줌 
code_df = code_df[['회사명', '종목코드']] # 우리가 필요한 것은 회사명과 종목코드이기 때문에 필요없는 column들은 제외해준다. 
code_df = code_df.rename(columns={'회사명': 'name', '종목코드': 'code'}) # 한글로된 컬럼명을 영어로 바꿔준다. 

print(code_df.head())

# 종목 이름을 입력하면 종목에 해당하는 코드를 불러와
# 네이버 금융(http://finance.naver.com)에 넣어줌

def get_url(item_name, code_df):
    code = code_df.query("name=='{}'".format(item_name))['code'].to_string(index=False).strip()
    url = 'http://finance.naver.com/item/sise_day.nhn?code={code}'.format(code=code)
    print("요청 URL = {}".format(url))

    return url

# GS의 일자데이터 url 가져오기

item_name='SK하이닉스'
url = get_url(item_name,code_df)

# 일자 데이터를 담을 df라는 DataFrame 정의
df = pd.DataFrame()

# 1페이지에서 20페이지의 데이터만 가져오기
for page in range(1, 125):
    pg_url = '{url}&page={page}'.format(url=url, page=page)
    df = df.append(pd.read_html(pg_url, header=0)[0], ignore_index=True)

df = df.dropna()
print(df.head())

df.to_csv('D:/Desktop/Itwill ws/rnn/hynix.csv', mode='w')

        name    code
0        DSR  155660
1         GS  078930
2      GS글로벌  001250
3  HDC현대산업개발  294870
4        KEC  092220
요청 URL = http://finance.naver.com/item/sise_day.nhn?code=000660
           날짜       종가     전일비       시가       고가       저가        거래량
1  2020.09.07  78400.0   300.0  78600.0  79500.0  77600.0  2730570.0
2  2020.09.04  78700.0     0.0  76400.0  79400.0  76100.0  4582566.0
3  2020.09.03  78700.0  3200.0  77700.0  79700.0  77200.0  7002529.0
4  2020.09.02  75500.0   300.0  75900.0  76400.0  74900.0  2369360.0
5  2020.09.01  75200.0   100.0  75200.0  76500.0  75200.0  2538676.0


### ※ 문제. 본인이 예측하기 원하는 회사의 주가를 기존에 했던 lstm 신경망에 넣고 똑같이 상관계수를 구하고 시각화도 하시오
    4년 데이터를 훈련 시키고 나머지 1년을 예측하시오

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.preprocessing import MinMaxScaler

# 1. 주가 데이터를 로드 합니다.
data = pd.read_csv("D:/Desktop/Itwill ws/rnn/hynix.csv")
print(data.tail())
print(data.shape)

data = data.rename(columns= {'날짜': 'date', '종가': 'close', '전일비': 'diff', '시가': 'open', '고가': 'high', '저가': 'low', '거래량': 'volume'})
data[['close', 'diff', 'open', 'high', 'low', 'volume']] = data[['close', 'diff', 'open', 'high', 'low', 'volume']].astype(int)  # 데이터의 타입을 int형으로 바꿔줌
data['date'] = pd.to_datetime(data['date'])
data = data.sort_values(by=['date'], ascending=False)
print(data.head())

# 2. 훈련, 테스트 데이터 나누기.
import datetime

data['date'] = pd.to_datetime(data['date'])
split_date = datetime.datetime(2019, 9, 5)
training_data = data[data['date'] < split_date].copy()
test_data = data[data['date'] >= split_date].copy()

scaler = MinMaxScaler()

data = data.drop(['Unnamed: 0', 'date','diff','volume'],axis=1)
data = scaler.fit_transform(data)
data_scale = scaler.scale_[0]
print(data_scale)

# 3. 예측을 위해 필요한 데이터만 남겨둡니다.
training_data = training_data.drop(['Unnamed: 0', 'date','diff','volume'], axis=1)
# training_data = np.array(training_data)

# 4. 정규화 합니다.
training_data = scaler.fit_transform(training_data)
print(training_data.shape)  # (991, 5)

# 5.
x_train = []
y_train = []

for i in range(60, training_data.shape[0]):
    x_train.append(training_data[i - 60:i])
    y_train.append(training_data[i, 0])

# 7. 훈련 데이터와 라벨을 numpy array 로 변환합니다.
x_train, y_train = np.array(x_train), np.array(y_train)

# 8. 신경망 모델을 생성합니다.
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense, LSTM, Dropout

regression = Sequential()
regression.add(LSTM(units=50, activation="relu", return_sequences=True, input_shape=(x_train.shape[1], 4)))
regression.add(Dropout(0.3))

regression.add(LSTM(units=60, activation="relu", return_sequences=True))
regression.add(Dropout(0.4))

regression.add(LSTM(units=80, activation="relu", return_sequences=True))
regression.add(Dropout(0.4))

regression.add(LSTM(units=120, activation="relu"))
regression.add(Dropout(0.5))

regression.add(Dense(units=1))

regression.summary()

# %% 9. 시각화 해봅니다.

# activate keras_study

# conda uninstall pydot
# conda uninstall pydotplus
# conda uninstall graphviz

# conda install pydot
# conda install pydotplus


# import tensorflow as tf

# tf.keras.utils.plot_model(regression, 'multi_input_and_output_model.png', show_shapes=True)

# 10. 학습
regression.compile(optimizer='adam', loss='mean_squared_error')
regression.fit(x_train, y_train, epochs=10, batch_size=32, verbose=0)

# 11.
data = pd.read_csv("D:/Desktop/Itwill ws/rnn/hynix.csv")
data = data.rename(columns= {'날짜': 'date', '종가': 'close', '전일비': 'diff', '시가': 'open', '고가': 'high', '저가': 'low', '거래량': 'volume'})
data[['close', 'diff', 'open', 'high', 'low', 'volume']] = data[['close', 'diff', 'open', 'high', 'low', 'volume']].astype(int)
data['date'] = pd.to_datetime(data['date'])
data = data.sort_values(by=['date'], ascending=False)
print(data.head())

data['date'] = pd.to_datetime(data['date'])
split_date = datetime.datetime(2019, 9, 5)
training_data = data[data['date'] < split_date].copy()
test_data = data[data['date'] >= split_date].copy()

past_60_days = training_data.head(60)

print(past_60_days.head())

df = test_data.append(past_60_days, ignore_index=True)
df = df.drop(['Unnamed: 0', 'date','diff','volume'], axis=1)

# test 데이터 정규화
testing_data = scaler.fit_transform(df)
print(testing_data.shape)

# df_ar = np.array(df)
# max(df['close'])

x_test = []
y_test = []

# testing_data = np.array(testing_data)

for i in range(60, testing_data.shape[0]):
    x_test.append(testing_data[i - 60:i])
    y_test.append(testing_data[i, 0])

x_test, y_test = np.array(x_test), np.array(y_test)
print(x_test.shape, y_test.shape)

# 예측
y_pred = regression.predict(x_test)
print(y_pred)

scale = 1 / data_scale
print(scale)
#
y_pred = y_pred * scale
y_test = y_test * scale
print(y_pred)
print(y_test)

# Visualising the results
# plt.figure(figsize=(14, 5))
plt.plot(y_test, color='red', label='Real Hynix Stock Price')
plt.plot(y_pred, color='blue', label='Predicted Hynix Stock Price')
plt.title('Hynix Stock Price Prediction')
plt.xlabel('Time')
plt.ylabel('Hynix Stock Price')
plt.legend()
plt.show()

from scipy import stats
import itertools
import numpy as  np

y_pred = np.array(y_pred)
y_pred = y_pred.flatten().tolist()
print(y_pred)
print(y_test)

print(np.corrcoef(y_pred, y_test))